### PR TITLE
Cardboard Support for Android

### PR DESCRIPTION
**Pull request by MrColdbird**
Long story short, virtual 3D cinema for PPSSPP.
Play your PSP games on a huge cinema screen without blowing out shit tons of money.

If you like the idea, just donate a few bucks to vanburace@gmail.com via PayPal.

### CODE REVIEW

**Comment by MrColdbird** [comment]
screenshot_2015-01-22-12-55-12
screenshot_2015-01-22-12-55-21

This commits have a few settings so this should work on every device, no matter the size or aspect ratio. So even Tablet users should have a chance to enjoy this.

**Comment by unknownbrackets** [comment]
Hmm, indentation seems to be mixed spaces in most cases.  Should change to tabs.

-[Unknown]

File: UI/GameSettingsScreen.cpp [hunk 1]
```diff
@@ -136,10 +136,21 @@ void GameSettingsScreen::CreateViews() {
...
+#ifdef ANDROID
+    graphicsSettings->Add(new ItemHeader(gs->T("Cardboard Settings", "Cardboard Settings")));
+    CheckBox *cardboardMode = graphicsSettings->Add(new CheckBox(&g_Config.bEnableCardboard, gs->T("Enable Cardboard", "Enable Cardboard")));
+    cardboardMode->SetDisabledPtr(&g_Config.bSoftwareRendering);
+	PopupSliderChoice * cardboardScreenSize = graphicsSettings->Add(new PopupSliderChoice(&g_Config.iCardboardScreenSize, 30, 100, gs->T("Cardboard Screen Size", "Screen Size (in % of the viewport)"), 1, screenManager()));
+    cardboardScreenSize->SetDisabledPtr(&g_Config.bSoftwareRendering);
+	PopupSliderChoice *cardboardXShift = graphicsSettings->Add(new PopupSliderChoice(&g_Config.iCardboardXShift, -100, 100, gs->T("Cardboard Screen X Shift", "X Shift (in % of the void)"), 1, screenManager()));
+    cardboardXShift->SetDisabledPtr(&g_Config.bSoftwareRendering);
+	PopupSliderChoice *cardboardYShift = graphicsSettings->Add(new PopupSliderChoice(&g_Config.iCardboardYShift, -100, 100, gs->T("Cardboard Screen Y Shift", "Y Shift (in % of the void)"), 1, screenManager()));
+    cardboardYShift->SetDisabledPtr(&g_Config.bSoftwareRendering);
```
Comment (unknownbrackets) [inline UI/GameSettingsScreen.cpp]: This should probably be enabled/disabled by g_Config.bEnableCardboard instead, or both that and software rendering.

Also, cool or not, I don't think Google Cardboard is so popular to warrant being at the very top of the settings screen.  Framerate and features most likely belong above it at the very least.

-[Unknown]

File: UI/GameSettingsScreen.cpp [hunk 2]
```diff
@@ -136,10 +136,21 @@ void GameSettingsScreen::CreateViews() {
...
 
+#ifdef ANDROID
+    graphicsSettings->Add(new ItemHeader(gs->T("Cardboard Settings", "Cardboard Settings")));
+    CheckBox *cardboardMode = graphicsSettings->Add(new CheckBox(&g_Config.bEnableCardboard, gs->T("Enable Cardboard", "Enable Cardboard")));
+    cardboardMode->SetDisabledPtr(&g_Config.bSoftwareRendering);
+	PopupSliderChoice * cardboardScreenSize = graphicsSettings->Add(new PopupSliderChoice(&g_Config.iCardboardScreenSize, 30, 100, gs->T("Cardboard Screen Size", "Screen Size (in % of the viewport)"), 1, screenManager()));
+    cardboardScreenSize->SetDisabledPtr(&g_Config.bSoftwareRendering);
+	PopupSliderChoice *cardboardXShift = graphicsSettings->Add(new PopupSliderChoice(&g_Config.iCardboardXShift, -100, 100, gs->T("Cardboard Screen X Shift", "X Shift (in % of the void)"), 1, screenManager()));
+    cardboardXShift->SetDisabledPtr(&g_Config.bSoftwareRendering);
+	PopupSliderChoice *cardboardYShift = graphicsSettings->Add(new PopupSliderChoice(&g_Config.iCardboardYShift, -100, 100, gs->T("Cardboard Screen Y Shift", "Y Shift (in % of the void)"), 1, screenManager()));
```
Comment (unknownbrackets) [inline UI/GameSettingsScreen.cpp]: I guess "the void" here means the empty / blank space outside the render area?  How do other Cardboard compatible games / applications present these options, do they also call it the "void"?

-[Unknown]

**Comment by unknownbrackets** [comment]
Hmm, is drawing the same image on both sides of the screen really enough to achieve the desired effect?  It doesn't need to be rendered from a shifted prespective or anything?

-[Unknown]

**Comment by hrydgard** [comment]
I've wanted to add an adjustable screen position in general (to not have to overlap touch buttons with the image on large tablets, without having to use the inflexible "small screen" option). Would be nice if the settings except a "cardboard" checkbox could be shared with that functionality (and thus named in a non-cardboard-specific way)

**Comment by MrColdbird** [comment]
Rendering the same image stereoscopically on both eye-viewports will result in a cinema-screen effect.

This means, the player will feel like playing the game on a 2D screen the size of about 20meter diameter.

The screen size parameter effectively decides how "close up" the player is sitting to the screen, making the screen itself seem bigger or smaller.

As for the terminology of "the void", Cardboard apps usually don't support changing parameters like this, why that is the case I don't know, but I found that giving it a Y-shift of about -20 works really well with centering the image on some cardboard models...

And I do believe that depending on what combination of device + cardboard you are using that tweaking of the parameters should be an option.

**Comment by hrydgard** [comment]
Yeah, for the best effect you need to distort the two rectangles a little too, but definitely not critical, this will look alright as long as you don't make the display extremely big.

Btw I'm not sure any longer if the parameters should be merged with some kind of general screen position/size, might be better to keep it by itself actually as it's really a standalone feature.

**Comment by MrColdbird** [comment]
I can tell you from trial and error with this commit of mine that the deformation due to lacking rectangle distortion is very minimal (below 20° on the corners) so I didn't even consider stretching to be a requirement for this kind of application.

Screen size settings from 30 (minimum I allowed) to 65% should fit in the viewport just fine without any distortion what so ever, beyond that the first bits of distortion show up but its so minimal that its no big deal.

For some games I also love playing at 100% viewport size, even if it cuts out periphery areas of the image. Games that work really well with this are for example Warriors Orochi 2, where your focus then lies entirely on your character, giving you a very awesome feeling when playing the game.

The normal users though will probably run this with the default settings I added to the commit, aka. around 50~60% screen size, resulting in the typical cinema experience with a fully visible view of the game.

**Comment by hrydgard** [comment]
Alright, sounds good. Fix the indentation and move it down a bit on the settings screen and I'll merge.

**Comment by MrColdbird** [comment]
There you go hrydgard. That should do it I hope.

**Pull request merged**